source/event_stream_channel_handler.c

```diff
@@ -0,0 +1,535 @@
+                "current message.",
+                (void *)handler,
+                remaining,
+                to_copy);
+
+            struct aws_byte_cursor to_append = aws_byte_cursor_advance(&message_cursor, to_copy);
+            if (aws_byte_buf_append_dynamic(&event_stream_handler->message_buf, &to_append)) {
+                error_code = aws_last_error();
+                AWS_LOGF_ERROR(
+                    AWS_LS_EVENT_STREAM_CHANNEL_HANDLER,
+                    "id=%p: Appending to the message buffer failed with error %s.",
+                    (void *)handler,
+                    aws_error_debug_str(error_code));
+
+                goto finished;
+            }
+        }
+
+        /* If we read the entire message, parse it and give it back to the subscriber. Keep in mind, once we're to this
+         * point the aws_event_stream API handles the rest of the message parsing and validation. */
+        if (event_stream_handler->message_buf.len == event_stream_handler->current_message_len) {
+            AWS_LOGF_TRACE(
+                AWS_LS_EVENT_STREAM_CHANNEL_HANDLER,
+                "id=%p: An entire message has been read. Parsing the message now.",
+                (void *)handler);
+            struct aws_event_stream_message received_message;
+            AWS_ZERO_STRUCT(received_message);
+
+            if (aws_event_stream_message_from_buffer(
+                    &received_message, event_stream_handler->handler.alloc, &event_stream_handler->message_buf)) {
+                error_code = aws_last_error();
+                AWS_LOGF_ERROR(
+                    AWS_LS_EVENT_STREAM_CHANNEL_HANDLER,
+                    "id=%p: Parsing the message failed with error %s.",
+                    (void *)handler,
+                    aws_error_debug_str(error_code));
+                goto finished;
+            }
+
+            size_t message_size = event_stream_handler->message_buf.len;
+            AWS_LOGF_TRACE(
+                AWS_LS_EVENT_STREAM_CHANNEL_HANDLER, "id=%p: Invoking on_message_received callback.", (void *)handler);
+            event_stream_handler->on_message_received(
+                &received_message, AWS_ERROR_SUCCESS, event_stream_handler->user_data);
+            aws_event_stream_message_clean_up(&received_message);
+            event_stream_handler->current_message_len = 0;
+            event_stream_handler->running_crc = 0;
+            aws_byte_buf_reset(&event_stream_handler->message_buf, true);
+
+            if (!event_stream_handler->manual_window_management) {
+                aws_channel_slot_increment_read_window(slot, message_size);
+            }
+        }
+    }
+
+finished:
+    if (error_code) {
+        event_stream_handler->on_message_received(NULL, error_code, event_stream_handler->user_data);
+        aws_channel_shutdown(slot->channel, error_code);
+    }
+    aws_mem_release(message->allocator, message);
+    return AWS_OP_SUCCESS;
+}
+
+struct message_write_data {
```

Comment:
not necessary now, but we should use the SBA for these objects, they are frequently allocated and tiny. Or maybe we should just put the SBA in front of the default allocator. All the task objects are also tiny and frequent.